src/python/pants/core/goals/tailor.py

```diff
@@ -582,10 +590,20 @@ async def tailor(
     union_membership: UnionMembership,
     specs: Specs,
     build_file_options: BuildFileOptions,
+    global_options: GlobalOptions,
 ) -> TailorGoal:
     tailor_subsystem.validate_build_file_name(build_file_options.patterns)
 
-    search_paths = PutativeTargetsSearchPaths(specs_to_dirs(specs))
+    if global_options.use_deprecated_directory_cli_args_semantics:
+        search_paths = PutativeTargetsSearchPaths(
+            dirs=(), deprecated_recursive_dirs=specs_to_dirs(specs)
+        )
+    else:
+        specs_paths = await Get(SpecsPaths, Specs, specs)
+        search_paths = PutativeTargetsSearchPaths(
+            tuple(sorted({os.path.dirname(f) for f in specs_paths.files}))
+        )
```

Comment:
We could theoretically get clever and try to set `recursive_dirs` sometimes...but this implementation is much simpler. Note that we can delete tailor's `specs_to_dirs` function once this deprecation cycle is over.